normalizer_lib.py

```diff
@@ -0,0 +1,158 @@
+
+        return union(TOKEN, EMAIL_ADDRESS, WEB_ADDRESS, TIME, FANCY_NUMBERS)
+
+
+# Step 1: Remove all extra whitespace between words
+
+REMOVE_EXTRA_WHITESPACE = cdrewrite(
+    transducer(SPACE, ""),
+    "",
+    SPACE,
+    SIGMA_STAR)
+
+
+# Step 2: Language-specific formatting fixes
+
+LANGUAGE_SPECIFIC_NORM = LANGUAGE.LANGUAGE_SPECIFIC_PREPROCESSING
+
+
+# Step 3: Apply NFC unicode normalize
+# really slow when done this way,
+# super fast when I just call unicodedata.normalize
+# on the string at the beginning
+# may remove later
+#def unicode_normalize(string: str) -> Fst:
+#    "Normalizes unicode to NFC normalization."
+#    nfc_normalized = unicodedata.normalize("NFC", string)
+#    normalizer = transducer(string, nfc_normalized).optimize()
+#    return cdrewrite(normalizer, "", "", SIGMA_STAR).optimize()
+
+
+# Step 4: Discard examples not associated with a pronunciation
+# e.g. "how are you today товарищ?" -> "how are you today?"
+
+
+def pass_only_valid_tokens(string: str) -> str:
+    """Replaces invalid tokens in a sentence with <UNK>, keeps the rest.
+
+    Args:
+        string: A line from a corpus.
+
+    Returns:
+        The line, but where invalid tokens have been replaced by <UNK>.
+    """
+    valid_token = INITIAL_PUNCTUATION.ques + Verbalizable.verbalizable() + FINAL_PUNCTUATION.ques
+    returned: List[str] = []
+    remove_extra_whitespace = (string @ REMOVE_EXTRA_WHITESPACE).string()
+    split_string = remove_extra_whitespace.split(" ")
+    for token in split_string:
+        if difference(acceptor(token), valid_token).num_states() == 0:
+            returned.append(token)
+        else:
+            returned.append("<UNK>")
+    return " ".join(returned)
+
+
+# Step 5: Detach punctuation from words
+# e.g. "Who are you?" -> "Who are you ?"
+
+
+NON_GRAPHEME_PUNCT = difference(PUNCTUATION, GRAPHEMES)
+
+INSERT_SPACE = transducer("", SPACE)
+
+SEPARATE_PUNCTUATION = (
+    cdrewrite(
+        INSERT_SPACE,
+        union("[BOS]", SPACE) + union(NON_GRAPHEME_PUNCT),
+        union(PUNCTUATION, Verbalizable.verbalizable()) + union("[EOS]", SPACE, Verbalizable.verbalizable()),
+        SIGMA_STAR) @
+    cdrewrite(
+        INSERT_SPACE,
+        union(Verbalizable.verbalizable(), PUNCTUATION),
+        PUNCTUATION + union("[EOS]", SPACE, PUNCTUATION),
+        SIGMA_STAR))
+
+
+# Step 7: Delete freestanding punctuation
+# e.g. "hi there ?" -> "hi there
+
+DELETE_PUNCTUATION = transducer(PUNCTUATION, "")
+
+DELETE_FREESTANDING_PUNCTUATION = cdrewrite(
+    DELETE_PUNCTUATION,
+    union("[BOS]", SPACE),
+    union("[EOS]", SPACE),
+    SIGMA_STAR)
+
+
+def normalizer(string: str) -> str:
+    """Applies FST rewrite rules to normalize text.
+
+    Args:
+        string: A line from a corpus.
+
+    Returns:
+        The normalized line from the corpus.
+    """
+    string = unicodedata.normalize("NFC", string.lower())
+    #filtered_string = string @ SENTENCE # for all-or-nothing filtering
```

Comment:
Instead of leaving this commented out, maybe make an alternative normalizer and add a test for it?

Reply:
That makes sense. Then we don't need to keep commenting different things in/out and can just call a different normalizer function if we want to filter out the whole sentence vs filtering out individual tokens.

Reply:
I went ahead and separated the normalizer function into two functions, one for token-based filtering and one for sentence-based filtering. In `normalizer_test.py` there are now separate tests for each one. In `normalizer.py`, if you give it a string to normalize it will call both functions so you can compare. If you don't give it a string (and it loads in a file instead), you can give it a flag telling it if you want to do token- or sentence-based filtering. It currently defaults to token-based.

One other thing to consider:  if we get a sentence that is e.g. 9 invalid tokens and 1 valid token, we might want to reject the whole sentence anyway. Maybe during the experiment phase we can explore different thresholds or methods for dealing with the fragments that result from replacing invalid tokens with `<UNK>`.